domain/schema/model.go

```diff
@@ -558,11 +561,87 @@ INSERT INTO storage_kind VALUES
     (0, 'block', 'Allows for the creation of raw storage volumes'), 
     (1, 'filesystem', 'Provides a hierarchical file storage system');
 
+-- This table stores storage constraint values for each named storage item
+-- defined by the application's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+CREATE TABLE application_storage_constraint (
+    application_uuid        TEXT NOT NULL,
+    storage_name            TEXT NOT NULL,
+    -- These attributes are filled in by sourcing data from:
+    -- user supplied, model config, charm config, opinionated fallbacks.
+    -- By the time the row is written, all values are known.
+    -- Constraint value types (pool, size, count) hitherto have
+    -- been fixed (since first implemented). We don't envisage
+    -- any change to how these are modelled.
+    --
+    -- Note: one might wonder why storage_pool below is not a
+    -- FK to a row defined in the storage pool table. This value
+    -- can also be one of the pool types. As with the comment on the
+    -- type column in the storage pool table, it's problematic to use a lookup
+    -- with an ID. Storage pools, once created, cannot be renamed so
+    -- this will not be able to become "orphaned".
+    storage_pool TEXT NOT NULL,
+    size         INT  NOT NULL,
+    count        INT  NOT NULL,
+    CONSTRAINT       fk_application_storage_constraint_application
+        FOREIGN KEY  (application_uuid)
+        REFERENCES   application(uuid),
+    PRIMARY KEY (application_uuid, storage_name)
+);
+
+-- Note that this is not unique; it speeds access by application.
+CREATE INDEX idx_application_storage_constraint
+ON application_storage_constraint (application_uuid);
+
+-- This table stores storage constraint values for each named storage item
+-- defined by the unit's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+-- Note: usually we just get the storage constraints off the application
+-- but need to allow for a unit's charm to temporarily diverge from that
+-- of its application.
+CREATE TABLE unit_storage_constraint (
+    unit_uuid               TEXT NOT NULL,
+    storage_name            TEXT NOT NULL,
```

Comment:
This looks like there should be a table for `charm_storage` with the names listed there and maintained on install/refresh.